microcosm_flask/conventions/swagger.py

```diff
@@ -0,0 +1,383 @@
+"""
+Swagger (OpenAPI) convention.
+
+Generates a Swagger definition, as JSON, for registered operations.
+
+Note that:
+ -  Swagger operations and type names use different conventions from the internal definitions
+    because we want to make usage friendly for code generation (e.g. bravado)
+
+ -  Marshmallow to JSON Schema conversion is somewhat simplistic. There are several projects
+    that already implement this conversion that we could try adapting. At the moment, the
+    overhead of adapting another library's conventions is too high.
+
+ -  All resource definitions are assumed to be shared and are declared in the "definitions"
+    section of the result.
+
+ -  All errors are defined generically.
+
+
+"""
+from flask import jsonify, g
+from inflection import camelize, pluralize
+from openapi import model as swagger
+from marshmallow import fields
+from werkzeug.routing import BuildError
+
+from microcosm.api import defaults
+from microcosm_flask.conventions.registry import (
+    get_qs_schema,
+    get_request_schema,
+    get_response_schema,
+    iter_operations,
+)
+from microcosm_flask.errors import ErrorSchema, ErrorContextSchema, SubErrorSchema
+from microcosm_flask.fields import EnumField
+from microcosm_flask.naming import name_for, singleton_path_for
+from microcosm_flask.operations import Operation
+
+
+def operation_name(operation, obj):
+    """
+    Convert an operation, obj(s) pair into a swagger operation id.
+
+    For compatability with Bravado, we want to use underscores instead of dots and
+    verb-friendly names. Example:
+
+        foo.retrieve       => client.foo.retrieve()
+        foo.search_for.bar => client.foo.search_for_bars()
+
+    """
+    if isinstance(obj, (list, tuple)):
+        verb, rest = operation.value.name, obj[1:]
+    else:
+        verb, rest = operation.value.name, []
+    return "_".join([verb] + [pluralize(name_for(noun)) for noun in rest])
+
+
+def type_name(name):
+    """
+    Convert an internal name into a swagger type name.
+
+    For example:
+
+        foo_bar => FooBar
+
+    """
+    if name.endswith("_schema"):
+        name = name[:-7]
+    return camelize(name)
+
+
+# see: https://github.com/marshmallow-code/apispec/blob/dev/apispec/ext/marshmallow/swagger.py
+FIELD_MAPPINGS = {
+    fields.Integer: ("integer", "int32"),
+    fields.Number: ("number", None),
+    fields.Float: ("number", "float"),
+    fields.Decimal: ("number", None),
+    fields.String: ("string", None),
+    fields.Boolean: ("boolean", None),
+    fields.UUID: ("string", "uuid"),
```

Comment:
One thing to note here: in swagger you have both types ("string") and formats ("uuid"). If you are using this particular format, it's not present in the Bravado client by default. Here's some code that adds it:

```
from uuid import UUID

def validate_uuid(s):
    try:
        UUID(s)
    except ValueError:
        raise SwaggerValidationError("Invalid UUID: {}".format(s))


uuid_format = SwaggerFormat(
    format="uuid",
    to_wire=lambda uuid: str(uuid),
    to_python=lambda s: s,
    validate=validate_uuid,
    description="UUID format",
)

config = {
    "formats": [uuid_format]
}

client = SwaggerClient.from_url(some_url, config=config)
```